### DISCUSSION TITLE
Middleware to inject EntityManager in context in NestJS Project

### DISCUSSION THREAD

**Discussion by NinjaKinshasa** [mikro-orm/mikro-orm#6304] [Q&A]
For testing purpose, our team decided to enable the reuse of the same EntityManager across multiple requests in our Nest project. To achieve this, we introduced a middleware that injects an EntityManager into the RequestContext. This way, any service that utilizes the EntityManager should resolve to the injected instance.

### Middleware Implementation

The following middleware is responsible for forking and injecting EntityManager:

```typescript
import { RequestContext } from '@mikro-orm/core';
import { Injectable, NestMiddleware } from '@nestjs/common';
import { NextFunction, Request, Response } from 'express';
import { EntityManager } from '@mikro-orm/core';

@Injectable()
export class TransactionMiddleware implements NestMiddleware {
  constructor(private readonly em: EntityManager) {}

  use(req: Request, res: Response, next: NextFunction) {
    // when header use-transaction is enabled, fork the current request context and inject it in the context with RequestContext.create
    if (req.headers['use-transaction'] === 'true') {
      const transaction = this.em.fork();
      console.log('Current request context em:', RequestContext.currentRequestContext()?.em?._id);
      transaction.begin();

      console.log('Injecting em:', transaction._id);

      RequestContext.create(transaction, () => {
        console.log('Current request context em after injection:', RequestContext.currentRequestContext()?.em?._id);
        next();
      });
    } else {
      next();
    }
  }
}

```
###  Controller Implementation

We implemented the following controller to verify the functionality of the injected EntityManager:

```typescript
import { Controller, Get } from '@nestjs/common';
import { EnsureRequestContext, EntityManager, RequestContext } from '@mikro-orm/core';

@Controller('demo')
export class DemoController {
  constructor(private readonly em: EntityManager) {}

  @Get()
  @EnsureRequestContext()
  async demo() {
    console.log('Em resolved in DemoController:', this.em._id);
    console.log('Current request context em in DemoController:', RequestContext.currentRequestContext()?.em?._id);
  }
}
```

### Unexpected Behavior

When calling the `/demo` endpoint using the following CURL command:

```bash
curl 'http://localhost:3000/demo' -H 'use-transaction: true
```

The logs show the following output:

```
Current request context em: undefined
Injecting em: 7
Current request context em after injection: 8
Em resolved in DemoController: 2
Current request context em in DemoController: 9
```

The entity manager resolved from the constructor of the service is not the injected one, and neither is the currentRequestContext EntityManager.

### Expected Behavior

What I expected to see instead was:
```
Current request context em: undefined
Injecting em: 7
Current request context em after injection: 7
Em resolved in DemoController: 7
Current request context em in DemoController: 7
```
### Questions

 - Why does RequestContext.create not persist the injected EntityManager ?
 - Why does the EntityManager resolved in the DemoController differ from the one injected in the middleware?
 - Why does the EntityManager resolved in the DemoController differ from the `RequestContext.currentRequestContext().em`?
 - How can I ensure that the EntityManager used across the request is consistent and matches the injected one?

I would greatly appreciate guidance on how to fix this issue or an explanation of what I might have misunderstood about the `RequestContext.create` method and its behavior in this scenario.

Repro demo : [https://github.com/NinjaKinshasa/mikro-orm-request-context](https://github.com/NinjaKinshasa/mikro-orm-request-context)

**Comment by B4nan** [comment]
`RequestContext.create` forks, you won't get the same EM with that.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/RequestContext.ts#L64

**Reply by NinjaKinshasa** [reply]
Thank you for your response.

> you won't get the same EM with that  

I was focusing on the `em._id`, but it actually does not matter whether it is the same `em` or not. What is important is that the code is executed within the same transaction started in the middleware, and that this transaction is not committed during the request.  

I updated the controller to explicitly roll back the transaction started in the controller:  

```typescript
import { Controller, Get } from '@nestjs/common';
import { EnsureRequestContext, EntityManager } from '@mikro-orm/core';

@Controller('demo')
export class DemoController {
  constructor(private readonly em: EntityManager) {}
  
  @Get()
  @EnsureRequestContext()
  async demo() {
    this.em.rollback();
  }
}
```

However, this results in an error:  

```
ValidationError: An open transaction is required for this operation
```

I expected the code to be executed within a transaction, but that does not seem to be the case.  

Additionally, if I attempt an insert operation:  

```typescript
import { Controller, Get } from '@nestjs/common';
import { EnsureRequestContext, EntityManager } from '@mikro-orm/core';
import { Author } from 'entities';

@Controller('demo')
export class DemoController {
  constructor(private readonly em: EntityManager) {}
  
  @Get()
  @EnsureRequestContext()
  async demo() {
    const user = this.em.create(Author, { name: 'John Doe', email: 'jdoe@example.com' });
    this.em.persistAndFlush(user);
  }
}
```

The author is inserted into the database. I expected it not to be, as the insert should have occurred within the transaction started in the middleware, and this transaction was not explicitly committed.

What I am trying to achieve is something like this : 

```
// 1. Fork and start transaction in the middleware --> database is empty
// 2. Perform tests that call controllers and insert into the database --> none of these changes should affect the database as they happen inside the transaction
// 3. End the test
// 4. Close the connection without committing the transaction --> database is still empty
```

Is it possible to do this ?

**Reply by B4nan** [reply]
IIRC there was someone trying similar approach for tests, not with a middleware though. There were some discussions on this topic, maybe in https://github.com/mikro-orm/nestjs, maybe here.

e.g. this one is similar https://github.com/mikro-orm/nestjs/discussions/107#discussioncomment-5169097
